### PR TITLE
🚀 feat(build): migrate to UV dependency management

### PR DESCRIPTION
**Pull request by samet-akcay**
## 📝 Description

- Migrated the build system from setuptools to hatchling for modern, standards-based packaging.
- Introduced a comprehensive `uv` configuration in `pyproject.toml` to manage PyTorch dependencies across multiple hardware backends (CPU, CUDA, ROCm, XPU).
- Refactored project dependencies into fine-grained optional-dependency groups, allowing for leaner installations.
- Updated all installation scripts and documentation (`README.md`, getting started guide) to use a clear, consistent hardware-first approach.
- Resolved dependency conflicts by pinning versions and using `uv`'s dependency resolver, ensuring a stable environment.

## ✨ Changes

Select what type of change your PR is:

- [x] 🚀 New feature (non-breaking change which adds functionality)
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 📦 Build system changes
- [x] 🔧 Chore (general maintenance)

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [x] 📚 I have made the necessary updates to the documentation (if applicable).
- [ ] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).
- [x] 🏷️ My PR title follows conventional commit format.

### Build System Migration to Hatchling

- Migrated from `setup.py` and `setuptools` to a `pyproject.toml`-based build powered by **hatchling**.
- The project version is now dynamically sourced from `src/anomalib/__init__.py`.

### `uv` Package Manager Integration

- A `[tool.uv]` section now manages the installation of PyTorch and its variants.
- **Custom PyTorch Indices**: Defines separate indices for each hardware backend to prevent dependency conflicts.
  - `pytorch-cpu`
  - `pytorch-cu118` (CUDA 11.8)
  - `pytorch-cu121` (CUDA 12.1)
  - `pytorch-cu124` (CUDA 12.4)
  - `pytorch-rocm61` (ROCm 6.1)
  - `pytorch-xpu` (Intel XPU)
- **Source Mapping**: The `[tool.uv.sources]` table directs `uv` to download `torch`, `torchvision`, and `pytorch-triton` from the correct index based on the selected installation extra.
- **Conflict Management**: The `[tool.uv.conflicts]` section prevents users from installing multiple hardware-specific extras simultaneously (e.g., `[cpu,cu124]`).

### Enhanced Optional Dependencies

- **Hardware-Specific Extras**: `torch` and `torchvision` have been moved from core dependencies to optional extras. Users must now explicitly choose a hardware backend, e.g., `anomalib[cpu]`.
- **Fine-Grained Model Dependencies**: Dependencies are organized into logical groups (`flow`, `feature`, `vlm_clip`, etc.) to keep installations minimal.
- **Convenience Groups**: `[full]` installs all model dependencies, and `[dev]` installs all testing and documentation tools.

## 🚀 Benefits

- **Reproducible Environments**: The `uv` configuration and lock file (`uv.lock`) ensure that all developers and CI systems use the exact same dependency versions.
- **Hardware-Optimized Installation**: Users can easily install a PyTorch build that is optimized for their specific hardware, improving performance.
- **Faster Dependency Resolution**: `uv` is significantly faster than `pip` and `conda` at resolving and installing packages.
- **Simplified Installation Workflow**: The installation process is now clearer, with a simple default and advanced options for power users.

#### Quick Install

For a standard installation. PyTorch will be installed based on its default behavior, which usually works for CPU and standard CUDA setups.

```bash
# With uv
uv pip install anomalib

# Or with pip
pip install anomalib
```

#### Advanced: Specifying Hardware & Features

Combine a hardware extra with any model extras you need.

```bash
# Install with CPU support
uv pip install "anomalib[cpu]"

# Install with CUDA 12.4 support
pip install "anomalib[cu124]"

# Install with ROCm and dependencies for flow-based models
uv pip install "anomalib[rocm,flow]"

# Install a full environment on CPU
pip install "anomalib[full,cpu]"
```

### For Developers (from Source)

```bash
git clone https://github.com/open-edge-platform/anomalib.git
cd anomalib

# Recommended: Use uv to create the environment from the lockfile
uv venv
uv sync --extra cpu  # For CPU

# Or for a different backend like CUDA
uv sync --extra cu124

# To set up a full development environment
uv sync --extra dev --extra cpu

# Alternative: Use pip for an editable install
pip install -e ".[dev,cpu]"
```

## 📊 Impact

- **Build System Modernization**: Migrated from legacy `setuptools` to modern, `pyproject.toml`-based `hatchling`.
- **Multi-Accelerator Support**: Added first-class support for PyTorch builds for CPU, CUDA (11.8, 12.1, 12.4), ROCm, and Intel XPU.
- **Improved DX**: Provides a faster, more reliable, and more flexible installation experience for both users and contributors.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -92,14 +97,115 @@ test = [
...
+    "torchvision>=0.19.0",
+]
+cu124 = [
+    "torch>=2.4.0",
+    "torchvision>=0.19.0",
+]
+rocm = [
+    "torch>=2.4.0",
+    "torchvision>=0.19.0",
+    "pytorch-triton-rocm ; sys_platform == 'linux'",
```
Comment (ashwinvaidya17) [inline pyproject.toml:120]: should this be darwin?

Comment (samet-akcay) [reply pyproject.toml:120]: no it should be linux as per the instructions 
https://docs.astral.sh/uv/guides/integration/pytorch/#using-a-pytorch-index

Comment (samet-akcay) [reply pyproject.toml:120]: I think mac does not need these extra indices

File: README.md [hunk 2]
```diff
@@ -61,31 +61,124 @@ Anomalib is a deep learning library that aims to collect state-of-the-art anomal
...
+uv pip install "anomalib[openvino,cu124]"
+
+# Example: Install all optional dependencies for a CPU-only setup
+uv pip install "anomalib[full,cpu]"
 ```
 
-## 🔧 Install from Source
+Here is a list of available optional dependency groups:
+
+| Extra         | Description                              | Models that need it                         |
```
Comment (rajeshgangireddy) [inline README.md]: Minor : 
"Models that need it" -> "Purpose" or "Use"

File: src/anomalib/models/components/filters/blur.py [hunk 3]
```diff
@@ -16,13 +16,24 @@
...
+
 import torch
-from kornia.filters import get_gaussian_kernel2d
-from kornia.filters.filter import _compute_padding
-from kornia.filters.kernels import normalize_kernel2d
+from lightning_utilities.core.imports import module_available
 from torch import nn
 from torch.nn import functional as F  # noqa: N812
 
+if TYPE_CHECKING or module_available("kornia"):
```
Comment (rajeshgangireddy) [inline src/anomalib/models/components/filters/blur.py]: why would kornia be not be installed. It is a core dependency right? Same with some other CV libraries (einops, freia)

**Review by rajeshgangireddy — COMMENTED** [review]
Minor comment:
In the readme file, should we update text on installing anomalib on intel GPUs in the section "# Training on Intel GPUs" (unless it's redundant) ?

**Review by rajeshgangireddy — APPROVED** [review]
